### PR TITLE
Update and delete note api

### PR DESCRIPTION
**Pull request by JamesCamel**
1. CodiMD note content update API 
  - PUT /api/notes/${noteId}
  - body: (JSON) content: string
2. CodiMD note delete API 
  - DELETE /api/notes/${noteId}
  - Check owenerId

### CODE REVIEW

File: package-lock.json [hunk 1]
```diff
@@ -7434,28 +7434,28 @@
       "dependencies": {
         "abbrev": {
           "version": "1.1.1",
-          "resolved": false,
+          "resolved": "https://registry.npmjs.org/abbrev/-/abbrev-1.1.1.tgz",
```
Comment (Yukaii) [inline package-lock.json:7437]: There's no changes in package.json so package-lock should not be commited.

File: lib/note/index.js [hunk 2]
```diff
@@ -232,7 +234,100 @@ function listMyNotes (req, res) {
...
+      if (realtime.isNoteExistsInPool(noteId)) {
+        logger.error('Update note failed: There are online users opening this note.')
+        return res.status('403').send({ status: 'error', message: 'Update API can only be used when no users is online' })
+      }
+
+      const now = Date.now()
+      const content = req.body.content
+      const updated = await note.update({
+        content: content,
+        lastchangeAt: moment(now).format('YYYY-MM-DD HH:mm:ss'),
```
Comment (Yukaii) [inline lib/note/index.js:294]: Update lastchangeAt with date directly is okay:

```suggestion
        lastchangeAt: now,
```

File: lib/note/index.js [hunk 3]
```diff
@@ -232,7 +234,100 @@ function listMyNotes (req, res) {
...
+      })
+
+      if (!updated) {
+        logger.error('Update note failed: Write note content error.')
+        return errorInternalError(req, res)
+      }
+
+      Revision.saveNoteRevision(note, (err, revision) => {
+        if (err) return errorInternalError(req, res)
+        if (!revision) return errorNotFound(req, res)
```
Comment (Yukaii) [inline lib/note/index.js:313]: Call `res.send` after revision is saved successfully (inside this callback)

File: lib/routes.js [hunk 4]
```diff
@@ -72,6 +72,10 @@ appRouter.get('/p/:shortid', response.showPublishSlide)
 appRouter.get('/p/:shortid/:action', response.publishSlideActions)
 // gey my note list
 appRouter.get('/api/notes/myNotes', noteController.listMyNotes)
+// delete note by id
+appRouter.delete('/api/notes/:noteId', noteController.deleteNote)
+// update note content by id
+appRouter.put('/api/notes/:noteId', urlencodedParser, noteController.updateNote)
```
Comment (Yukaii) [inline lib/routes.js:78]: Use json parser instead

```suggestion
appRouter.put('/api/notes/:noteId', bodyparser.json(), noteController.updateNote)
```

File: lib/note/index.js [hunk 5]
```diff
@@ -232,7 +234,100 @@ function listMyNotes (req, res) {
...
+        ]
+      })
+
+      if (!updated) {
+        logger.error('Update note failed: Write note content error.')
+        return errorInternalError(req, res)
+      }
+
+      Revision.saveNoteRevision(note, (err, revision) => {
+        if (err) return errorInternalError(req, res)
```
Comment (Yukaii) [inline lib/note/index.js:312]:
```javascript
logger.error(err)
```

File: lib/note/index.js [hunk 6]
```diff
@@ -232,7 +234,100 @@ function listMyNotes (req, res) {
...
+        if (!revision) return errorNotFound(req, res)
+      })
+
+      res.send({
+        status: 'ok'
+      })
+    } catch (err) {
+      logger.error('Update note failed: Internal Error.')
```
Comment (Yukaii) [inline lib/note/index.js:320]:
```javascript
logger.error(err)
```

Comment (JamesCamel) [reply lib/note/index.js:322]: Fix.

File: lib/note/index.js (hunk 5 above)
Comment (JamesCamel) [reply lib/note/index.js]: Fix.

File: lib/routes.js (hunk 4 above)
Comment (JamesCamel) [reply lib/routes.js]: Fix.

File: lib/note/index.js (hunk 3 above)
Comment (JamesCamel) [reply lib/note/index.js:315]: Fix.

File: lib/note/index.js (hunk 2 above)
Comment (JamesCamel) [reply lib/note/index.js]: Fix.

File: lib/note/index.js [hunk 7]
```diff
@@ -232,7 +233,103 @@ function listMyNotes (req, res) {
...
+      }
+
+      if (realtime.isNoteExistsInPool(noteId)) {
+        logger.error('Update note failed: There are online users opening this note.')
+        return res.status('403').json({ status: 'error', message: 'Update API can only be used when no users is online' })
+      }
+
+      const now = Date.now()
+      const content = req.body.content
+      const updated = await note.update({
```
Comment (Yukaii) [inline lib/note/index.js:291]: The title should also be updated:

```diff
+ title: Note.parseNoteTitle(content)
```

File: lib/note/index.js [hunk 8]
```diff
@@ -232,7 +233,103 @@ function listMyNotes (req, res) {
...
+            now
+          ]
+        ]
+      })
+
+      if (!updated) {
+        logger.error('Update note failed: Write note content error.')
+        return errorInternalError(req, res)
+      }
+
```
Comment (Yukaii) [inline lib/note/index.js:309]: Update user history like:

https://github.com/hackmdio/codimd/blob/091c77bdb1f05263543f85f5c6f04f5bfaf3c722/lib/note/index.js#L47

**Comment by Yukaii** [comment]
Thank you @JamesCamel

**Pull request merged**